idaes/apps/grid_integration/multiperiod/design_and_operation_models.py

```diff
@@ -0,0 +1,121 @@
+#################################################################################
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES).
+#
+# Copyright (c) 2018-2023 by the software owners: The Regents of the
+# University of California, through Lawrence Berkeley National Laboratory,
+# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
+from functools import reduce
+from pyomo.environ import (
+    Var,
+    Param,
+    Binary,
+    Expression,
+    NonNegativeReals,
+    Constraint,
+)
+from pyomo.common.config import ConfigValue, In
+from idaes.core.base.process_base import declare_process_block_class
+from idaes.core.base.process_base import ProcessBlockData
+
+
+@declare_process_block_class("DesignModel")
+class DesignModelData(ProcessBlockData):
+    """
+    Class for containing design model ...
```

Comment:
I still do not know enough about what this class actually does to approve this. The docs do not explain what this is or how to use it, and I am not sure what this code actually adds (i.e. is it even necessary?).

For one, this needs a descriptive doc string to explain what it does.

Reply:
@radhakrishnatg would be the best person to respond to this.

Reply:
Based on discussions, @radhakrishnatg explained the need for these classes, particularly with regard to cost calculations later on. @andrewlee94 are there other particulars that you want addressed, aside from further documentation?

Reply:
@radhakrishnatg and I have discussed this. The first big thing is some basic documentation so that users and maintainers can understand what they are looking at (this can just be better doc strings). I also pointed out that the same basic interface could be done using a more standard approach of having the users write derived classes instead of the current approach of providing a method as an argument (it adds 2 or 3 lines of boilerplate code, but aligns this tool with all the rest of IDAES).

I would also like to make sure there is a maintenance plan for this and an assigned maintainer.

Reply:
@andrewlee94 Do you mean that we would have Design and Operation model classes as is, with the model build config options (and associated code) removed, then build model specific classes? For example, NGCC design model would inherit from DesignModel class, and it would merely add the required code under the build method to construct the model; similarly, NGCC operation model would inherit from OperationModel and add required code under build method. Is that along the lines of what you were thinking?

In the interim, we would be maintaining this code for flex desal work, but I imagine that there are other activities that would make use of this code as well, now and in the future. I guess the assigned maintainer(s) would be up for discussion. Arguably, it could be @radhakrishnatg , @MarcusHolly , and myself.

Reply:
Yes, that was my suggestion as it is almost identical to writing the method that is passed in as an argument.

My main reason is because this is how the other IDAES models are written and so aligns this with the standard. Otherwise we have this one section of the code that does things differently which will end up confusing someone at some point.

For the transition, you could support both - someone writing the build method just does not provide the model build config. However, you would want to add a deprecation to the `model_build` pathway and not document it (which is easy as there is no documentation yet).

Reply:
I am pasting my response here for better visibility:

@andrewlee94 I was just thinking about this, and on the one hand, using derived classes would be more in line with how the rest of IDAES does things. On the other hand, using this model_func approach gives the potential of applying either approach: I can formulate quick and dirty models on the fly by using the config option to pass in a build function, or I can go and create a child class. So in a sense, allowing for the current approach would give more flexibility, while eliminating it would be more restrictive on the user (which might be the way to go in most cases, but not necessarily here).

In any case, I think that any refactor of this formulation should be part of a subsequent PR. I think it is more important for this overall capability to get merged in first, since there has already been published work on this Pricetaker approach (for one).

Reply:
@adam-a-a My counter to the argument against derived classes is that from a code perspective there are only two additional lines of boilerplate code, and thus derived classes are effectively no harder that the `model_build` approach. Consider:

```
def my_func():
    # My code
```

is equivalent to:

```
@declare_process_block_class(MyPriceTaker)
class MyPriceTakerData(PriceTakerBase):
    def build():
        # My code
```

The one difference is how arguments to the method would be handled, but even that is fairly easy. For the derived class approach, the code just needs to refer to `self.config.model_build_arguments` (which already exists) instead of expecting the mas direct arguments.

I do agree that this could be done in a later PR however. The main reason this came up was that it is not particularly well documented how this was intended to be used so I wanted clarification on useage for this PR.